File: code/Java/src/main/java/Infrastructure/SpaceComm/MarsRoverSender.java
package Infrastructure.SpaceComm;

import Infrastructure.Bus.ISendFinalStateBus;
import Infrastructure.ISendNotifications;

public class MarsRoverSender implements ISendNotifications {

    private INasaAntenna nasaAntenna;

    public MarsRoverSender(INasaAntenna nasaAntenna) {
        this.nasaAntenna = nasaAntenna;
    }

    public void readsFrom(ISendFinalStateBus marsRoverServiceBus) {
        marsRoverServiceBus.trigger(this);
    }

    public void send(String message) {
        String[] messageParts = message.split(" ");
        nasaAntenna.received(new String[]{
                "X" + messageParts[0],
                "Y" + messageParts[1],
                "D" + messageParts[2]
        });
    }

    public void sendError() {
        nasaAntenna.received(new String[]{"ER"});
    }
}

File: code/CSharp/ConnascentMarsRover/Source/Infrastructure/SpaceComm/MarsRoverSender.cs
using System;
using Source.Infrastructure.Bus;

namespace Source.Infrastructure.SpaceComm
{
    public class MarsRoverSender : ISendNotifications
    {
        private readonly INasaAntenna _nasaAntenna;

        public MarsRoverSender(INasaAntenna nasaAntenna)
        {
            _nasaAntenna = nasaAntenna;
        }

        public void Send(string message)
        {
            var messageParts = message.Split(" ");
            _nasaAntenna.Received(new[]
            {
                $"X{messageParts[0]}",
                $"Y{messageParts[1]}",
                $"D{messageParts[2]}"
            });
        }

        public void ReadsFrom(ServiceBus marsRoverServiceBus)
        {
            marsRoverServiceBus.Trigger(this);
        }

        public void SendError()
        {
            _nasaAntenna.Received(new[] {"ER"});
        }
    }
}